Add block C

fnand opened the pull request:
Hi,

So I have been attempting to write block C as specified in 1007:3300.

So far I have not been completely successful, and this should probably not be pulled into master, but into a new side branch. 

I am not sure whether my implementation is  faulty, or the values that I am using to test it result in ridiculous values (i.e. GIGO). 

I have added a new example as a test, as well as a new test in tests/phaseSpaceVolume/blockC

I have been working from blockD as a template. 
See https://github.com/fnand/MoMEMta/blob/add-blockC/BlockC2_eq.nb and/or https://github.com/fnand/MoMEMta/blob/add-blockC/BlockC.pdf for the algebra. 

It's possible a made a logic error somewhere along the way, or even a typo, but as far as I understand this is the way it should be done. Hopefully someone can spot my mistake. 

-Ferdi

swertz commented:
Thanks a lot for this! You even included a unit test :+1:

Something I'd ask right away is for you to clean the pull request to keep only these three files:
- modules/BlockC.cc
- tests/unit_tests/modules.cc
- CMakeLists.txt

You can keep the rest for you in another branch of course. 
We're going to reorganise the tests soon (cross section & phase-space volume) so we'd like to avoid adding more stuff that's going to be changed anyway.
Of course if there's other things - not related to block C - that you would like to get included, you're welcome to open another PR!

Once that is done and the project compiles (for now the integration test fails) I'll pull your branch locally and check the block itself ;)

fnand commented:
Hi, 

I removed all the unnecessary changes, as well as the unit test I originally put in modules.cc

The test was basically just a skeleton and failed, meaning travis complained. My real test was the one in the phaseSpaces which I removed at your request.

swertz commented:
Great, thanks, I'll have a more in-depth look now :+1:.

Actually I think Travis failed because of some other reason... You're welcome to put the unit test back into `modules.cc`: from what I see the test itself is fine and should succeed if block C is correct!

fnand commented:
Hi, thanks, removed the latest commit then. 
The test is somewhat slapdash as it uses the input values from the other test (for blockD). 

The test without it failed with an error about the ttbar matrix element, which i am pretty sure I didn't mess around with.

swertz commented on tests/unit_tests/modules.cc:
```diff
@@ -229,4 +229,53 @@ TEST_CASE("Modules", "[modules]") {
+        auto module = createModule("BlockC");
+
+        REQUIRE(module->work() == Module::Status::OK);
+        REQUIRE(solutions->size() == 2);
```
The rest of the test looks good; this is the only line that might have to be changed since there might be up to four solutions, and for the particular inputs used here the number of solutions has to be retrieved at least once from the block.

swertz commented on core/src/MoMEMta.cc:
```diff
@@ -169,7 +169,7 @@ std::vector<std::pair<double, double>> MoMEMta::computeWeights(const std::vector
 
     *m_met = met;
-
+	
```
Please remove this

swertz commented on cmake/modules/FindLHAPDF.cmake:
```diff
@@ -17,8 +17,8 @@ if (NOT LHAPDF_INCLUDE_DIR OR NOT LHAPDF_LIBRARY)
-        exec_program(${LHAPDF_CONFIG_PROGRAM} ARGS "--incdir" OUTPUT_VARIABLE LHAPDF_INCLUDE_DIR)
-        exec_program(${LHAPDF_CONFIG_PROGRAM} ARGS "--libdir" OUTPUT_VARIABLE LHAPDF_LIBRARY_DIR)
+        exec_program(${LHAPDF_CONFIG_PROGRAM} ARGS "--incdir"  OUTPUT_VARIABLE LHAPDF_INCLUDE_DIR)
+        exec_program(${LHAPDF_CONFIG_PROGRAM} ARGS "--libdir"  OUTPUT_VARIABLE LHAPDF_LIBRARY_DIR)
```
Please revert these changes

swertz commented on CMakeLists.txt:
```diff
@@ -297,6 +298,7 @@ if(EXAMPLES)
     set_target_properties(example_WW_fullyleptonic PROPERTIES OUTPUT_NAME
       "WW_fullyleptonic.exe")
 
+
```
Please remove the empty line

swertz commented on tests/unit_tests/modules.cc:
```diff
@@ -229,4 +229,53 @@ TEST_CASE("Modules", "[modules]") {
+            REQUIRE(solution.valid == true);
+
+            LorentzVector test_p12  = input_particles->at(0) + solution.values.at(0);
+            LorentzVector test_p123 = input_particles->at(1) + test_p12;
```
Again, for `p3` you shouldn't use the input one but the block's output, ie `solution.values.at(1)` once you've added it.

swertz commented:
I think the expressions would get shorter and less error-prone if you defined `X=p2*p3/E3` since it appears in several places ;) Also computing it once and for all would make things quicker.

fnand commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,370 @@
+                pT = - *m_met;
+            } else {
+		//Err, do I need to add that other quark
+                pT = p2 + p3;
```
I'm not certain how this should work. How do the branches work?

swertz replied:
So, the branches is all the other reconstructed/visible objects in the event, beside p1 and p3 which are the ones that get reconstructed by this block.
In this case this means only p2, plus the extra branches, which could be extra particles or could come from secondary blocks.

fnand replied:
Great thanks!

fnand commented:
Hi, so I made some of the changes you suggested. I will more carefully look at tabs alignment soon, but my primary concern is with the fact that something doesn't work. 

One of the strange things is I often get messages similar to:

> No solution to equation 0 x^2 + 0 x + -5.96046e-08

other than that the s12 calculated is usually pretty far from the given value, which is an obvious clue that something is wrong. 

I will go through the equations again, but yes, CForm sucks and I eventually didn't use the CForm, just the normal output. 

Any idea what is wrong with my implementation, or how I can check it granularly?

fnand commented:
Some examples of outputs I get for calulated s12 vs given s12 is: 

`Calculated:  298761  Input   275480
Calculated:  201579  Input   247018
Calculated:  199799  Input   247018
Calculated:  212706  Input   233726
Calculated:  -266845  Input   318164
Calculated:  -266845  Input   318164
Calculated:  383650  Input   318164
Calculated:  357671  Input   245967
Calculated:  249490  Input   307394`

Which are significant differences

swertz commented:
I have found the problem :)

So the function `solve2Quads` doesn't work when `a11=0`, which is the case here. I'm sorry, this should be documented. You can simply solve it here: https://github.com/MoMEMta/MoMEMta/pull/137/files#diff-5d8ea00c7c37f24044bef758f6f78750R213 by renaming a<->b from lines 213-226.

Actually when defining the alphas (https://github.com/MoMEMta/MoMEMta/pull/137/files#diff-5d8ea00c7c37f24044bef758f6f78750R183 and so on), since some of them are zero, you can simply throw them away, and this will simplify the expressions for aXX and bXX...

swertz commented:
Ok I'll let you fix the remaining issues and the style, and then we'll have a final round ;)

fnand commented:
Perfect yeah that solves it. 

Will do some cleaning up and then commit later

fnand commented:
Done, see if this is ok?

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+
+            m1 = parameters.get<double>("m1", 0.);
+
+            m_particles.push_back(get<LorentzVector>(parameters.get<InputTag>("p2")));
```
Since there are only two we might avoid using a vector, which also avoids the step https://github.com/MoMEMta/MoMEMta/pull/137/files#diff-5d8ea00c7c37f24044bef758f6f78750R125? @blinkseb an opinion?

blinkseb replied:
Indeed please avoid the `vector` here, no need to be dynamic since you know in advance you only have 2 particles. Just declare `p2` and `p3` directly as member of the class.

fnand replied:
Makes sense. Done.

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+            // one ("met::p4")
+            InputTag met_tag;
+            if (parameters.exists("met")) {
+
```
Please remove empty line

fnand replied:
Done

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+            // p1y = ...(2)
+            // p1z = ...(3)
+            // E3  = ...(4)            
+            const double p2Sq   = p2.Dot(p2);
```
Please use `p2.M2()`

fnand replied:
Done

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+            const double denom = 2.*(E2 - p2z*costhe3 - p2x*cosphi3*sinthe3 - p2y*sinthe3*sinphi3 );
+                  
+            // Term appears regularly, compute once.    
+            const double X = p2x*sinthe3*cosphi3 + p2y*sinthe3*sinphi3 + p2z*costhe3;
```
Since it appears in `denom` it could be used there as well?

And actually we also have `X=p2.Vect().Dot(p3.Vect())/p3.P()` or `denom = 2*p2.Dot(p3)/p3.P()`... Not sure which method is quicker or more readable in the end, I'll leave it up to you.

fnand replied:
I switched around `X` and `denom` and used `X` in `denom`. 
I did not change it to the vector implementation, do you think it would be faster that way? It has a div in there so I am not sure, but I haven't tested it.

fnand replied:
I guess if I was really concerned with optimization I could calculate 1/denom and then multiply by it...

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+                                   + p2y*sinthe3*sinphi3*(SQ(m1) + p2Sq + 2*p2x*pTx - *s123))/(-p2z*denom);
+
+            const double beta4  = -1./denom;
+            const double gamma4 = ( *s123- *s12)/denom;
```
Please leave no space after an opening parenthesis and before a closing parenthesis (also in the lines below).

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+            if (E1.size() == 0)
+                return Status::NEXT;
+
+            for(unsigned int i=0; i<E1.size(); i++){
```
A few spaces missing -> `for (std::size_t i = 0; i < E1.size(); i++) {`

fnand replied:
Done. But would you prefer `std::size_t`? All the other blocks use `unsigned int`. It's probably just a matter of consistency in this case as we don't expect large amounts of solutions.

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+                // Make sure E3 is not negative 
+                if (e3  < 0.) 
+                    continue;
+
```
Please only one empty line (also below)

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+
+                double jacobian = computeJacobian(p1, p2, p3);
+                
+                Solution s {{p1, p3}, jacobian, true };
```
Either add a space after the first `{` or remove the one after the last `}` ;)

fnand replied:
Done

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+            return solutions->size() > 0 ? Status::OK : Status::NEXT;
+        }
+
+        double computeJacobian(const LorentzVector& p1, const LorentzVector& p2, const LorentzVector& p3) {
```
In my previous round of comment I suggested to move all this part in the loop above, to avoid re-accessing all the vector components that have already been computed, ie. define `p1x = beta1*alp + gamma1`, and so on, and use them both to define `p1` and `p2` and compute the jacobian.

fnand replied:
I must have missed that one. The reason I did it this way is because this is how it is done in BlockD. 

Maybe I am misunderstanding the logic, but wouldn't moving the return statement exit the loop prematurely?

fnand replied:
Ah, you mean move the computeJacobian into the loop?

swertz replied:
Yes exactly as in there ;) https://github.com/MoMEMta/MoMEMta/blob/master/modules/BlockE.cc#L248

swertz commented on tests/unit_tests/modules.cc:
```diff
@@ -229,4 +229,53 @@ TEST_CASE("Modules", "[modules]") {
+            REQUIRE(test_p123.M2() == Approx(s_123));
+
+ //         LorentzVector test_pT = test_p134 + test_p256;
+ //         REQUIRE(test_pT.Pt() == Approx(0));
```
You can uncomment (and adapt) the Pt() test ;)

You can also require that the input and output masses of `p1` and `p3` are "equal".

swertz commented on tests/unit_tests/modules.cc:
```diff
@@ -229,4 +229,53 @@ TEST_CASE("Modules", "[modules]") {
+    }
+
+
+
```
Please max. one empty line

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+/*
+ *  MoMEMta: a modular implementation of the Matrix Element Method
+ *  Copyright (C) 2016  Universite catholique de Louvain (UCL), Belgium
```
2017

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+#include <momemta/Types.h>
+#include <momemta/Math.h>
+
+/** \brief \f$\require{cancel}\f$ Final (main) Block C, describing \f$q_1 q_2 \to \f$X + s_{123} (\to p_3 + s_{12} (\to \cancel{p_1} p_2))\f$
```
Please remove `q_1 q_2 \to \f$X + ` as they are not present in the final parameterisation

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+ *
+ * This Block addresses the change of variables needed to pass from the standard phase-space
+ * parametrisation for \f$p_{1 \dots 3} \times \delta^4\f$ to a parametrisation in terms of the two (squared) masses
+ * of the intermediate propagators.
```
... and the angular variables of `p3`

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+ * parametrisation for \f$p_{1 \dots 3} \times \delta^4\f$ to a parametrisation in terms of the two (squared) masses
+ * of the intermediate propagators.
+ *
+ * The integration is performed over \f$s_{12}, s_{123}\f$ with \f$p_{1 \dots 3}\f$ as input. Per integration point,
```
same here, please mention `theta_3` and `phi_3`

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+ * of the intermediate propagators.
+ *
+ * The integration is performed over \f$s_{12}, s_{123}\f$ with \f$p_{1 \dots 3}\f$ as input. Per integration point,
+ * the LorentzVector of the invisible particle, \f$p_1\f$, is computed based on the following set
```
... and also `p3`'s energy
Maybe that's the also right place to mention that `p3` is required to be massless.

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+ *   | Name | Type | %Description |
+ *   |------|------|--------------|
+ *   | `s12` <br/> `s123` <br/>  | double | Squared invariant masses of the propagators. Typically coming from a BreitWignerGenerator or NarrowWidthApproximation module. |
+ *   | `p1` ... `p3` | LorentzVector | LorentzVectors of the particles used to reconstruct the event according to the above method. |
```
`p2` is not required as input, so you can replace the dots by `<br/>`. Also please mention that for `p3` only angles and mass are used, as in https://github.com/MoMEMta/MoMEMta/blob/master/modules/BlockA.cc#L62

blinkseb replied:
Actually from the code inputs are `p2` and `p3`.

swertz replied:
You're right, that was a typo ;)

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,303 @@
+ *
+ *   | Name | Type | %Description |
+ *   |------|------|--------------|
+ *   | `solutions` | vector(Solution) | Solutions of the change of variable. Each solution embed  the LorentzVectors of the invisible particle (ie. \f$(p_1)\f$) and the associated jacobian. These solutions should be fed as input to the Looper module. |
```
`Each solution embeds`
Also, `p3` should be mentioned here

fnand commented:
Hi, 

Please check if the documentation is ok now. I wrote it before changing anything and subsequently forgot to update it. 

Also, the unit test is not yet complete. 

Right now I am just hijacking the values that were specified for block D. 
What would the suggested way to handle the inputs for BlockC be?

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,296 @@
+ * parametrisation for \f$p_{1 \dots 3} \times \delta^4\f$ to a parametrisation in terms of the two (squared) masses
+ * of the intermediate propagators and the angular variables of \f$p3\f$.
+ *
+ * The integration is performed over \f$s_{12}, s_{123}\, \theta_3, \phi_3f$ with \f$p_{1 \dots 3}\f$ as input. Per integration point,
```
There's a `\` missing before the `$`

swertz commented on modules/BlockC.cc:
```diff
@@ -243,9 +236,9 @@ class BlockC: public Module {
-                double jacobian = computeJacobian(p1, p2, p3);
-                
-                Solution s {{p1, p3}, jacobian, true };
+                double jacobian = computeJacobian(p1, *p2, p3);
```
I would call the `p3` defined L223 differently, e.g. `p3_sol` or something like that. It works because you're in the loop, but it's not the same `p3` as above so it's not ideal for readability...

swertz commented:
Thanks! For the unit test, it's perfectly OK to re-use the same inputs as for block D! I think the test just fails because `p3` has to be massless for block C, and the "random" 4-vectors we're using to test block D are massive. So what you can do is add an extra "random" massless 4-vector, in the same way as it's done for the others, and use that one as `p3` input to test block C ;)

fnand commented:
I think all the issues except the unit test have been addressed now. Anything that I missed?

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,274 @@
+                if(q1Pz > sqrt_s/2 || q2Pz > sqrt_s/2)
+                    continue;
+
+                const double p1x = p1.Px();
```
Instead of retrieving those here, you can define `p1x = beta1*alp + gamma1`, ..., and then use those both to define `p1` and `p3_sol` and to compute the jacobian.

swertz commented:
Thanks! I only had one small comment. For the unit test there's really not much missing ;)

blinkseb commented:
Thanks a lot for your contribution! Before we can merge it, you'll need to rebase your PR in top of `master`, and also squash your changes into a single commit, maybe two if you want to keep the unit tests separated from the main block. It'd be also nice if you can add a line to the change log, stating the block C was added!

Thanks again! :+1:

swertz commented:
Btw if you rebase on master you'll be able to use one of https://github.com/MoMEMta/MoMEMta/blob/master/tests/unit_tests/modules.cc#L81 as input for `p3` to test block C ;)

fnand commented:
Hi, 

So rebase has been completed. Commits have all been squashed into one.

Everything is still not fine with the unit test. Which either means the values I am giving it are unreasonable, or something is fishy in block c. 

I get   6394.8875476811 == Approx( 6400.0 ) currently. 

Also, is it just me or do I have to use input 0, otherwise the module won't accept it?

blinkseb commented on CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
  - Secondary blocks A, B, CD, E
  - Main blocks A, E and G (not present in MadWeight)
  - `getRandom4Vector` function to generate random Lorentz vectors of a specified mass (useful in cases where a particle has to be passed from C++, but integrated over all its components).
+ - Main Block C and unit test for it
```
Please just edit line 25 to add `C` (no need for the unit test in the change log)

fnand replied:
Done

blinkseb commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,274 @@
+ *
+ *   | Name | Type | %Description |
+ *   |------|------|--------------|
+ *   | `pT_is_met` | bool, default false | Fix \f$\vec{p}_{T}^{tot} = -\vec{\cancel{E_T}}\f$ or \f$\vec{p}_{T}^{tot} = \sum_{i \in \text{ vis}} \vec{p}_i\f$ |
```
I see that your block accepts `m1` as a parameter but it's not documented. Can you add a line?

fnand replied:
Done

blinkseb commented on CHANGELOG.md:
```diff
@@ -24,7 +24,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
  - Main blocks A, E and G (not present in MadWeight)
  - `getRandom4Vector` function to generate random Lorentz vectors of a specified mass (useful in cases where a particle has to be passed from C++, but integrated over all its components).
- - Main Block C and unit test for it
+ - Main Block C
```
I meant to remove this line altogether and edit line 25 to read `Main blocks A, C, E and G (not present in MadWeight)`

blinkseb commented on modules/BlockC.cc:
```diff
@@ -65,7 +65,7 @@
  *   | `branches` | vector(LorentzVector) | LorentzVectors of all the other particles in the event, taken into account when computing \f$\vec{p}_{T}^{branches}\f$ (if MET is not used), and checking if the solutions are physical. |
  *   | `met` | LorentzVector, default `met::p4` | LorentzVector of the MET |
- *
+ *   | `m1`  | Mass of the invisible particle, default 1 | 
```
Please respect the structure of the table, there're 3 column, `name`, `type` and `description`. Default mass is 0, not 1.

```
| `m1` | double, default to 0 | Mass of the invisible particle | 
```

And it's not an input, but a parameter, please move the line to the correct section of the documentation.

fnand replied:
Ah, great, and I also wrote the incorrect default value. Fixing it now

swertz commented:
Hi Ferdinand,

After checking extensively on our side, it does look like something's broken again with the block's solutions...

If I remember well, it did work perfectly at some point, so I'm guessing one of the changes you've implemented has broken it, unless you did check regularly that the solutions were correct (which would have been a good thing to do)? 

Unfortunately we can't access either your commit history or the mathematica files to check where the problem might be...

Would you mind checking again the different expressions? I think it has to be somewhere  between here https://github.com/MoMEMta/MoMEMta/pull/137/files#diff-5d8ea00c7c37f24044bef758f6f78750R157 and here https://github.com/MoMEMta/MoMEMta/pull/137/files#diff-5d8ea00c7c37f24044bef758f6f78750R178, the rest looks fine to me.

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,274 @@
+            const double beta2  = (sinthe3*sinphi3)/denom;
+            const double gamma2 = -(2*E2*pTy - 2*X*pTy - *s12*sinthe3*sinphi3 + *s123*sinthe3*sinphi3)/denom;
+
+            const double alpha3 = 2*(-SQ(E2) + E2*X)/(-p2z*denom);
```
`alpha3 = E2 / p2z`

swertz commented on modules/BlockC.cc:
```diff
@@ -0,0 +1,274 @@
+
+            const double alpha3 = 2*(-SQ(E2) + E2*X)/(-p2z*denom);
+            const double beta3  = (p2x*cosphi3*sinthe3 + p2y*sinthe3*sinphi3)/(-p2z*denom);
+            const double gamma3 = ((-E2 + p2z*costhe3)*(SQ(m1) + p2Sq + 2*p2x*pTx + 2*p2y*pTy - *s12)
```
`gamma3 = 0.5 * (- *s12 + SQ(m1) + p2Sq + 2 * p2x * (pTx + sinthe3 * cosphi3 * (*s123 - *s12) / denom) + 2 * p2y * (pTy + sinthe3 * sinphi3 * (*s123 - *s12) / denom) ) / p2z`

fnand replied:
This solved it. I am trying to check where I went wrong. I suspect a typo in copying from mathematica.

swertz commented:
Found the problem, see the two comments.

Now the only thing that's missing in the unit test is to require that the input and output masses for `p1` are equal, and that `p3.M / p3.E = 0`.

fnand commented:
Thanks a lot. I am trying to figure out what went wrong. I suspect an error when copying from mathematica. I added the check for m1 before and after, as well as `p3.M  = 0`.
What do you mean by `p3.M / p3.E = 0`.

blinkseb commented:
Awesome :+1: ! Once we are all in favor for merging, I'll run `clang-format` once to reformat the code (there's a lot of spaces missing around) and then push the green button.

Thanks again!

swertz commented:
Great thanks, the tests are now passing :+1:

I meant doing like that: https://github.com/MoMEMta/MoMEMta/blob/master/tests/unit_tests/modules.cc#L246
I think it makes more sense to check that the mass is zero with a precision relative to the particle's energy (when `E` and `p` are large, computing `E^2 - p^2` will never be precisely zero)...?

swertz commented:
:+1: for me, @blinkseb it's all yours ;)

blinkseb commented:
Thanks a lot! I merged this PR manually, see f44d80242e9cb6f62fd5da0369f445517c063842.
